### ISSUE TITLE
Storing data/coding shred data in rocksdb causes problems

### ISSUE THREAD

**Issue by sakridge** [solana-labs/solana#16234]
#### Problem

https://github.com/solana-labs/solana/issues/14586
* High write IO from write amplification
* Inability to delete individual dead slots without bad performance
* Corruption on restart

#### Proposed Solution

Replace rocksdb with accounts_db

**Comment by steviez** [comment]
For bookkeeping, link to a previous attempt at solving this problem
https://github.com/solana-labs/solana/pull/9366

**Comment by sakridge** [comment]
If you have 200m shreds there are about 650,000 slots. AccountsDB uses around 200 bytes per index in memory. 200m * 200 bytes is 40 gigabytes of memory just for the indexes.

If you index per-slot, then you have about 300x less indexing, maybe 150mb total which isn't bad, but there will be extra overhead of serializing/re-serializing and extracting the item from the slot if you need to read individual shreds. Also extra read/write bandwidth to access them.

Being able to page-out unused indexes is an option and would probably help the accounts use-case as well.

On the other hand, custom indexing using the file system itself to bucket shreds into slots may be even more efficient. This would be something like the previous solution #9366.

**Comment by steviez** [comment]
Interesting project I stumbled across; not all these are applicable to the approach we may be taking, but still useful to review what other people have already done.
https://github.com/bytedance/terarkdb

> The main reasons TerarkDB works better [than RocksDB] in some use cases are:
> - TerarkDB fixed tons of bugs in the original RocksDB.
> - TerarkDB moved lots of code out from db mutex, especially those critical code paths.
> - TerarkDB optimized background I/O patterns a lot to reduce disk stress.
> - TerarkDB optimized WAL sync strategy to reduce user thread latency.
> - TerarkDB has a better KV separation implementation.
> - TerarkDB has a lazy compaction mechanism to reduce write amplification A LOT (on heavy write workloads)
> - TerarkDB has bunch of self-developed indexing and compression algorithms that is more efficient than traditional ways.
> - ......

**Comment by steviez** [comment]
Notes from a call with @carllin, I'll type these items a bit formally but gotta run right now (FYI @sakridge)
- Do we care about minor forks? Do we care about keeping these around or can we chuck them?
- Have background flush; the trigger for flushing cache
   - Roots being set OR memory size exceeding some pre-determined
   - If we drop shreds but can recover with erasure, those will be filled in and can complete a slot (if possible)
   - Partial slots might not be flushed, how do we know when to evict these?
- Iterators: handling the case where a slot may be flushed while iteration is on-going
   - Similar structure to accountsDB: `DashMap<Slot, Arc<DashMap<u64, Shred>>>`

**Comment by ryoqun** [comment]
For the record, i think this is no longer true:

>     * Corruption on restart

I think this is referring to two cases:

- `Corruption: block checksum mismatch`: i think this is actually due to hardware problem as the upstream bug report says: (details: https://github.com/solana-labs/solana/issues/9009#issuecomment-896879129)
- `Corruption: SST file is ahead of WALs`: this should have been fixed upstream (details: https://github.com/solana-labs/solana/pull/10838#issuecomment-887178263)

Hope these hard-earned progress could result in better informed decision for the usage of continued rocksdb usage for rpc, bankless leader's block cost, etc, even after we move data/coding shred.

(to be fair, I'm neutral to the idea of ditching rocksdb for shreds. I'm thinking tweaking rocksdb might be easier than re-inventing wheel.... but there is no time to prove it by actual code/tuning in near time...)

**Comment by sakridge** [comment]
> For the record, i think this is no longer true:
> Corruption: block checksum mismatch: i think this is actually due to hardware problem as the upstream bug report says: (details: Rocksdb corruption issues on compaction #9009 (comment))

Yes. This does seem to be less of a problem now, although we did encounter it on another machine a couple months ago.

One aspect of that is that it would be nice to simply recover the database even if we have this single bit corruption. The validator can throw away the data which is corrupt and retrieve it from the network again or from the coding shreds. I'm not sure how easy or hard rocksdb makes this. I've read that the accepted method is to open the DB in read-only mode and then copy it to a whole new DB. That doesn't sound like it would be very fast for a 500gb DB. Possibly we can reverse-engineer the rocks data format to get what we want, but it's probably subject to change unless they are open to some modification. Or maybe there is a way already that I don't know about. If we control the data format on-disk, it would be pretty straightforward to do though.

**Comment by sakridge** [comment]
Too many RPC reads from ledger data stored in rocksdb seems to cause the validator to fall behind as well.

**Comment by ryoqun** [comment]
> Too many RPC reads from ledger data stored in rocksdb seems to cause the validator to fall behind as well.

i'll detail later. but it seems that the previously-found rayon hack (`SOLANA_RAYON_THREADS=8`) and the below hack alleviate the problem:

```
$ diff -U3 /home/ryoqun/.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-epoch-0.9.{3,5}/src/internal.rs
--- /home/ryoqun/.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-epoch-0.9.3/src/internal.rs  1970-01-01 09:00:00.000000000 +0900
+++ /home/ryoqun/.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-epoch-0.9.5/src/internal.rs  2021-12-11 08:18:54.633505937 +0900
@@ -387,7 +387,7 @@
 impl Local {
     /// Number of pinnings after which a participant will execute some deferred functions from the
     /// global queue.
-    const PINNINGS_BETWEEN_COLLECT: usize = 128;
+    const PINNINGS_BETWEEN_COLLECT: usize = 128 * 128;
 
     /// Registers a new `Local` in the provided `Global`.
     pub(crate) fn register(collector: &Collector) -> LocalHandle {

```
